File: KMHManagedObject.m
//
//  KMHManagedObject.m
//  KMHCoreDataController
//
//  Created by Ken M. Haggerty on 4/13/16.
//  Copyright © 2016 Ken M. Haggerty. All rights reserved.
//

#pragma mark - // NOTES (Private) //

#pragma mark - // IMPORTS (Private) //

#import "KMHManagedObject.h"

#pragma mark - // DEFINITIONS (Private) //

NSString * const KMHManagedObjectNotificationObjectKey = @"object";

NSString * const KMHManagedObjectWillBeDeallocatedNotification = @"kNotificationKMHManagedObject_WillBeDeallocated";
NSString * const KMHManagedObjectWasCreatedNotification = @"kNotificationKMHManagedObject_WasCreated";
NSString * const KMHManagedObjectWasFetchedNotification = @"kNotificationKMHManagedObject_WasFetched";
NSString * const KMHManagedObjectWillSaveNotification = @"kNotificationKMHManagedObject_WillSave";
NSString * const KMHManagedObjectDidSaveNotification = @"kNotificationKMHManagedObject_DidSave";
NSString * const KMHManagedObjectWillBeDeletedNotification = @"kNotificationKMHManagedObject_WillBeDeleted";

@interface KMHManagedObject ()
@property (nonatomic, strong, readwrite) NSSet *changedKeys;
@property (nonatomic, readwrite) BOOL isSaving;
@property (nonatomic, readwrite) BOOL willBeDeleted;
@property (nonatomic, readwrite) BOOL wasDeleted;
@end

@implementation KMHManagedObject

#pragma mark - // SETTERS AND GETTERS //

@synthesize changedKeys = _changedKeys;
@synthesize isSaving = _isSaving;
@synthesize willBeDeleted = _willBeDeleted;
@synthesize wasDeleted = _wasDeleted;
@synthesize parentIsDeleted = _parentIsDeleted;

#pragma mark - // INITS AND LOADS //

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] postNotificationName:KMHManagedObjectWillBeDeallocatedNotification object:self userInfo:nil];
}

- (void)awakeFromInsert {
    [super awakeFromInsert];
    
    [[NSNotificationCenter defaultCenter] postNotificationName:KMHManagedObjectWasCreatedNotification object:self userInfo:nil];
}

- (void)awakeFromFetch {
    [super awakeFromFetch];
    
    [[NSNotificationCenter defaultCenter] postNotificationName:KMHManagedObjectWasFetchedNotification object:self userInfo:nil];
}

- (void)willSave {
    self.isSaving = YES;
    
    [super willSave];
    
    if (self.isUpdated && !self.isInserted) {
        self.changedKeys = [NSMutableSet setWithArray:self.changedValues.allKeys];
    }
    
    NSDictionary *userInfo = self.changedKeys ? @{KMHManagedObjectNotificationObjectKey : self.changedKeys} : @{};
    [[NSNotificationCenter defaultCenter] postNotificationName:KMHManagedObjectWillSaveNotification object:self userInfo:userInfo];
}

- (void)didSave {
    if (self.willBeDeleted) {
        self.willBeDeleted = NO;
        self.wasDeleted = YES;
    }
    
    if (self.changedKeys && !self.inserted) { // !self.isDeleted &&
        NSDictionary *userInfo = @{KMHManagedObjectNotificationObjectKey : self.changedKeys};
        [[NSNotificationCenter defaultCenter] postNotificationName:KMHManagedObjectDidSaveNotification object:self userInfo:userInfo];
        self.changedKeys = nil;
    }
    
    [super didSave];
    
    self.isSaving = NO;
}

- (void)prepareForDeletion {
    self.willBeDeleted = YES;
    
    [[NSNotificationCenter defaultCenter] postNotificationName:KMHManagedObjectWillBeDeletedNotification object:self userInfo:nil];
    
    [super prepareForDeletion];
}

#pragma mark - // PUBLIC METHODS //

- (void)setup {
    self.isSaving = NO;
    self.wasDeleted = NO;
    self.parentIsDeleted = NO;
}

#pragma mark - // CATEGORY METHODS //

#pragma mark - // DELEGATED METHODS //

#pragma mark - // OVERWRITTEN METHODS //

#pragma mark - // PRIVATE METHODS //

@end

File: KMHManagedObject.swift
//
//  KMHManagedObject.swift
//  KMHCoreDataController
//
//  Created by Ken M. Haggerty on 6/14/16.
//  Copyright © 2016 Peter Cicchino Youth Project. All rights reserved.
//

// MARK: Imports

import Foundation
import CoreData

// MARK: Protocols

// MARK: Definitions

let KMHManagedObjectNotificationObjectKey = "object"

let KMHManagedObjectWillBeDeallocatedNotification = NSNotification.Name("kKMHManagedObjectWillBeDeallocatedNotification")
let KMHManagedObjectWasCreatedNotification = NSNotification.Name("kKMHManagedObjectWasCreatedNotification")
let KMHManagedObjectWasFetchedNotification = NSNotification.Name("kKMHManagedObjectWasFetchedNotification")
let KMHManagedObjectWillSaveNotification = NSNotification.Name("kKMHManagedObjectWillSaveNotification")
let KMHManagedObjectDidSaveNotification = NSNotification.Name("kKMHManagedObjectDidSaveNotification")
let KMHManagedObjectWillBeDeletedNotification = NSNotification.Name("kKMHManagedObjectWillBeDeletedNotification")

// MARK: - KMHManagedObject

class KMHManagedObject: NSManagedObject {
    
    // MARK: - Public API
    
    // MARK: • IBOutlets
    
    // MARK: • Variables
    
    private(set) var changedKeys: [String]?
    private(set) var isSaving: Bool = false
    private(set) var willBeDeleted: Bool = false
    private(set) var wasDeleted: Bool = false
    private(set) var parentIsDeleted: Bool = false
    
    // MARK: • Functions
    
    // MARK: - Private API
    
    // MARK: • IBOutlets
    
    // MARK: • Variables
    
    // MARK: • Inits
    
    deinit {
        NotificationCenter().post(name: KMHManagedObjectWillBeDeallocatedNotification, object: self, userInfo: nil)
    }
    
    override init(entity: NSEntityDescription, insertInto context: NSManagedObjectContext?) {
        super.init(entity: entity, insertInto: context)
    }
    
    override func awakeFromInsert() {
        super.awakeFromInsert()
        
        NotificationCenter().post(name: KMHManagedObjectWasCreatedNotification, object: self, userInfo: nil)
    }
    
    override func awakeFromFetch() {
        super.awakeFromFetch()
        
        NotificationCenter().post(name: KMHManagedObjectWasFetchedNotification, object: self, userInfo: nil)
    }
    
    override func willSave() {
        self.isSaving = true
        
        super.willSave()
        
        if (self.isUpdated && !self.isInserted) {
            self.changedKeys = Array(self.changedValues().keys)
        }
        
        NotificationCenter().post(name: KMHManagedObjectWillSaveNotification, object: self, userInfo: [KMHManagedObjectNotificationObjectKey : self.changedKeys])
    }
    
    override func didSave() {
        if (self.willBeDeleted) {
            self.willBeDeleted = false
            self.wasDeleted = true
        }
        
        if ((self.changedKeys != nil) && !self.isInserted) { // && !self.isDeleted
            NotificationCenter().post(name: KMHManagedObjectDidSaveNotification, object: self, userInfo: [KMHManagedObjectNotificationObjectKey : self.changedKeys!])
            self.changedKeys = nil
        }
        
        super.didSave()
        
        self.isSaving = false
    }
    
    // MARK: • Setup
    
    // MARK: • Delegated
    
    // MARK: • Functions

}
